rclrs/src/wait/guard_condition.rs

```diff
@@ -0,0 +1,173 @@
+/// by adding an `Arc<GuardCondition>` to the wait set, and calling `trigger()` on the same `GuardCondition` while
+/// the wait set is waiting.
+///
+/// The guard condition may be reused multiple times, but like other waitable entities, can not be used in
+/// multiple wait sets concurrently.
+///
+/// # Example
+/// ```
+/// # use rclrs::{Context, GuardCondition, WaitSet, RclrsError};
+/// # use std::sync::{Arc, atomic::Ordering};
+///
+/// let context = Context::new([])?;
+///
+/// let atomic_bool = Arc::new(std::sync::atomic::AtomicBool::new(false));
+/// let atomic_bool_for_closure = Arc::clone(&atomic_bool);
+///
+/// let gc = Arc::new(GuardCondition::new(
+///     &context,
+///     Some(Box::new(move || {
+///         atomic_bool_for_closure.store(true, Ordering::Relaxed);
+///     })),
+/// ));
+///
+/// let mut ws = WaitSet::new(0, 1, 0, 0, 0, 0, &context)?;
+/// ws.add_guard_condition(Arc::clone(&gc))?;
+///
+/// // Trigger the guard condition, firing the callback and waking the wait set being waited on, if any.
+/// gc.trigger()?;
+///
+/// // The provided callback has now been called.
+/// assert_eq!(atomic_bool.load(Ordering::Relaxed), true);
+///
+/// // The wait call will now immediately return.
+/// ws.wait(Some(std::time::Duration::from_millis(10)))?;
+///
+/// # Ok::<(), RclrsError>(())
+/// ```
+pub struct GuardCondition {
+    /// The rcl_guard_condition_t that this struct encapsulates.
+    pub(crate) rcl_guard_condition: Mutex<rcl_guard_condition_t>,
+    /// An optional callback to call when this guard condition is triggered.
+    callback: Option<Box<dyn Fn() + Send + Sync>>,
+    /// A flag to indicate if this guard condition has already been assigned to a wait set.
+    pub(crate) in_use_by_wait_set: Arc<AtomicBool>,
+}
+
+impl Drop for GuardCondition {
+    fn drop(&mut self) {
+        unsafe {
+            // SAFETY: No precondition for this function (besides passing in a valid guard condition)
+            rcl_guard_condition_fini(&mut *self.rcl_guard_condition.lock().unwrap());
+        }
+    }
+}
+
+impl PartialEq for GuardCondition {
+    fn eq(&self, other: &Self) -> bool {
+        // Because GuardCondition controls the creation of the rcl_guard_condition, each unique GuardCondition should have a unique
+        // rcl_guard_condition. Thus comparing equality of this member should be enough.
+        std::ptr::eq(
+            &self.rcl_guard_condition.lock().unwrap().impl_,
+            &other.rcl_guard_condition.lock().unwrap().impl_,
+        )
+    }
+}
+
+impl Eq for GuardCondition {}
+
+// SAFETY: rcl_guard_condition is the only member that doesn't implement Send, and it is designed to be accessed from other threads
+unsafe impl Send for rcl_guard_condition_t {}
+
+impl GuardCondition {
+    /// Creates a new guard condition.
+    pub fn new(context: &Context, callback: Option<Box<dyn Fn() + Send + Sync>>) -> Self {
```

Comment:
For slightly better usability, we could make this take an un-boxed callback and box it ourselves in the function body.

Reply:
Hrm. I played around with this a bit, and couldn't figure out how to get around the `Sized` requirement without the `Box`. Any tips?

Reply:
You can write
```
    pub fn new<F>(context: &Context, callback: Option<F>) -> Self
    where
        F: Fn() + Send + Sync + 'static,
```
and below
```
        Self {
            rcl_guard_condition: Mutex::new(guard_condition),
            callback: callback.map(|cb| Box::new(cb) as Box<dyn Fn() + Send + Sync>),
            in_use_by_wait_set: Arc::new(AtomicBool::new(false)),
        }
```

Reply:
Pushed! I thought I tried this but I think I tried to cast inside the box instead of outside, which obviously wouldn't work.